### PR TITLE
Add the IntroText and mobile homepage search bar

### LINKED ISSUE #511: Search bar should be front and center on home
I think we should add the search bar back front and center into the home page. Especially for mobile where it's not visible at all. For project pages it still make sense to just have it in the title bar. 

v1:
signal-2022-11-24-175209

v2:
signal-2022-11-24-175130_002

### PR DESCRIPTION
**Pull request by AbdulrhmnGhanem**
- Refactor: `SearchContext.jsx` -> `SearchContext.tsx`
- UI: improve `mobile_menu` breakpoint
- UI: make the search bar usable in homepage too
- UI: add search bar in home page for small devices
- e2e: `Homepage search on mobile`

##
Fixes #511.

### CODE REVIEW

**Comment by kasbah** [comment]
Some currently open PR is causing the `+` in the "add a project" button to be grey:

image

I'm not sure if it's this one or another one, but this one is touching the bar styles and is using grey...

**Comment by kasbah** [comment]
I think the search input should remain in the results page on mobile as well.

**Comment by AbdulrhmnGhanem** [comment]
> I think the search input should remain in the results page on mobile as well.

I am not sure what you mean.

**Comment by kasbah** [comment]
I feel the experience is currently a little confusing on mobile. You search and then the search input seems to disappear. It goes into the hamburger menu but as a user you don't know thst.

**Comment by AbdulrhmnGhanem** [comment]
The search is visible ` /search` too. Could you add a screenshot of the page?
image

**Comment by kasbah** [comment]
Ah, something weird is going on when there are no results. 

Screenshot_20230111-234314

**Comment by kasbah** [comment]
Also the "Kitspace is..." blurb isn't going away like it does on v1.

**Comment by AbdulrhmnGhanem** [comment]
The copy and search field are visible now even if no search results exist.

image

I handled icon color regression too.
|regression|current|
|---|---|
|image|image|

**Comment by kasbah** [comment]
The copy should disappear though.

**Comment by kasbah** [comment]
Now there's not enough of a gap between the search and results on mobile. 

image

File: frontend/src/components/ProjectCardGrid/index.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ const ProjectCardGrid = ({
   intersectionObserverRef,
 }: ProjectCardGridProps) => {
   // Set the marker to load more projects to 6 cards before the end of the grid.
-  const loadMoreProjectsMarker = projects.length - 6
+  const loadMoreProjectsMarker = projects?.length - 6
```
Comment (kasbah) [inline frontend/src/components/ProjectCardGrid/index.tsx:16]: Will `loadMoreProjectsMarker` be `NaN` or `-6` when projects is `undefined`? How about when it's `null`?

Seems like a situation where it should be an empty array if there's no projects and never `null` or `undefined`.

File: frontend/src/pages/search/index.tsx [hunk 2]
```diff
@@ -55,22 +56,37 @@ const Search = ({ swrFallback, initialQuery }: SearchPageProps) => {
...
-  }
 
   return (
-    <ProjectCardGrid
-      intersectionObserverRef={intersectionObserverRef}
-      projects={projects}
-    />
+    <>
+      <SearchBar className={styles.mobileSearchBar} />
+      {/*  Show the copy only when the query is empty */}
```
Comment (kasbah) [inline frontend/src/pages/search/index.tsx:63]: This comment feels a little useless. 

```suggestion
```

File: frontend/test/Copy.test.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import Search from '@pages/search'
+
+import CopyFixture from './fixtures/Copy.json'
+
+beforeAll(() => {
+  vi.mock('next/config', () => ({
+    default: () => ({
+      publicRuntimeConfig: {
+        KITSPACE_GITEA_URL: 'https://gitea.kitspace.test',
+        KITSPACE_ASSET_URL: 'https://s3.kitspace.test',
```
Comment (kasbah) [inline frontend/test/Copy.test.tsx:12]: We don't currently use a `s3.kitspace.test` domain.

File: frontend/test/fixtures/Copy.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+  "swrFallback": {
+    "$inf$#query:\"\",offset:0,limit:18,": [
```
Comment (kasbah) [inline frontend/test/fixtures/Copy.json:3]: Doesn't seem like a good idea to have the cache key in a fixture. Does the test actually need to check the `swrFallback`? Does a user care what's in the `swrFallback`?

**Review by kasbah — CHANGES_REQUESTED** [review]
It's a definite improvement so willing to merge as is with minor corrections. I think there are still some usability issues with mobile search but will open new issues for them.

File: frontend/src/components/ProjectCardGrid/index.tsx (hunk 1 above)
Comment (AbdulrhmnGhanem) [reply frontend/src/components/ProjectCardGrid/index.tsx:16]: Resolved in 098c1853.

File: frontend/src/pages/search/index.tsx (hunk 2 above)
Comment (AbdulrhmnGhanem) [reply frontend/src/pages/search/index.tsx]: Resolved in 098c1853.

File: frontend/test/Copy.test.tsx (hunk 3 above)
Comment (AbdulrhmnGhanem) [reply frontend/test/Copy.test.tsx]: Resolved in 2c0fcbb9.

File: frontend/test/fixtures/Copy.json (hunk 4 above)
Comment (AbdulrhmnGhanem) [reply frontend/test/fixtures/Copy.json]: Resolved in 2c0fcbb9.

**Comment by kasbah** [comment]
Can we disambiguate "Copy" and call it "IntroText" or someting?

**Comment by AbdulrhmnGhanem** [comment]
> Can we disambiguate "Copy" and call it "IntroText" or something?

Changed the title, and component name in f5155ed0.

**Pull request merged**